components/camel-kubernetes/src/main/java/org/apache/camel/component/kubernetes/customresources/KubernetesCustomResourcesProducer.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.camel.support.DefaultProducer;
+import org.apache.camel.support.MessageHelper;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.camel.util.json.JsonArray;
+import org.apache.camel.util.json.JsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KubernetesCustomResourcesProducer extends DefaultProducer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KubernetesCustomResourcesProducer.class);
+
+    public KubernetesCustomResourcesProducer(AbstractKubernetesEndpoint endpoint) {
+        super(endpoint);
+    }
+
+    @Override
+    public AbstractKubernetesEndpoint getEndpoint() {
+        return (AbstractKubernetesEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    public void process(Exchange exchange) throws Exception {
+        String operation;
+
+        if (ObjectHelper.isEmpty(getEndpoint().getKubernetesConfiguration().getOperation())) {
+            operation = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_OPERATION, String.class);
+        } else {
+            operation = getEndpoint().getKubernetesConfiguration().getOperation();
+        }
+
+        switch (operation) {
+
+            case KubernetesOperations.LIST_CUSTOMRESOURCES:
+                doList(exchange, operation);
+                break;
+
+            case KubernetesOperations.LIST_CUSTOMRESOURCES_BY_LABELS_OPERATION:
+                doListByLabels(exchange, operation);
+                break;
+
+            case KubernetesOperations.GET_CUSTOMRESOURCE:
+                doGet(exchange, operation);
+                break;
+
+            case KubernetesOperations.DELETE_CUSTOMRESOURCE:
+                doDelete(exchange, operation);
+                break;
+
+            case KubernetesOperations.CREATE_CUSTOMRESOURCE:
+                doCreate(exchange, operation);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Unsupported operation " + operation);
+        }
+    }
+
+    protected void doList(Exchange exchange, String operation) throws Exception {
+        String namespaceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_NAMESPACE_NAME, String.class);
+        JsonObject customResourcesListJSON = new JsonObject(
+                getEndpoint().getKubernetesClient().customResource(getCRDContext(exchange.getIn())).list(namespaceName));
+        LOG.info(customResourcesListJSON.toString());
+        JsonArray customResourcesListItems = new JsonArray(customResourcesListJSON.getCollection("items"));
+
+        MessageHelper.copyHeaders(exchange.getIn(), exchange.getOut(), true);
+        exchange.getOut().setBody(customResourcesListItems);
+    }
+
+    protected void doListByLabels(Exchange exchange, String operation) throws Exception {
+        String namespaceName = exchange.getIn().getHeader(KubernetesConstants.KUBERNETES_NAMESPACE_NAME, String.class);
```

Comment:
Apart from my other comment, I think these configurations can be obtained either from the endpoint configurations or the exchange headers, not only the exchange headers, isn't? If so, IMHO, I'd add a logic to first check if these options being set in the exchange headers first and fall back to the endpoint options aka `KubernetesConfiguration`.

Reply:
Thanks!  I'll add it.